### PR TITLE
[Syntax] Clean up syntax grammar for declaration nodes

### PR DESCRIPTION
**Pull request by DougGregor**
Collected cleanups for the declaration grammar used by SwiftSyntax/libSyntax:
* Split `ActorDecl` into its own syntax node, separate from `ClassDecl`
* Update the list of declaration modifiers to match that of the language
* Use `FunctionSignature` for initializers rather than its own parameter clause, which picks up (e.g.) proper support for async and the experimental reasync, and also unifies their parameter lists with that of function declarations
* Factored out the "detail" of a decl modifier, so we have an optional parenthesized detail list rather than optional components
* `TypeAliasDecl`'s initializer (the `= type`) is not optional, so remove its optionality

### CODE REVIEW

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/499

@swift-ci please clean test

File: test/Syntax/round_trip_parse_gen.swift [hunk 1]
```diff
@@ -148,7 +148,6 @@ typealias J = inout @autoclosure () -> Int
 typealias K = (@invalidAttr Int, inout Int, __shared Int, __owned Int) -> ()
 
 @objc private typealias T<a,b> = Int
-@objc private typealias T<a,b>
```
Comment (ahoppen) [inline test/Syntax/round_trip_parse_gen.swift:151]: Why did you need to remove this test case?

Comment (DougGregor) [reply test/Syntax/round_trip_parse_gen.swift:151]: Because a `typealias` always needs to have an `= <type>` to be syntactically well-formed, and this test requires that the input source be syntactically well-formed to pass. This is no longer representable without an "unknown declaration" error due to the change to make the `= type` required by the syntax grammar.

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/499
https://github.com/apple/swift-stress-tester/pull/180

@swift-ci please clean test

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/499
https://github.com/apple/swift-stress-tester/pull/180

@swift-ci please clean test

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/499
https://github.com/apple/swift-stress-tester/pull/180

@swift-ci please clean test

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/499
https://github.com/apple/swift-stress-tester/pull/180

@swift-ci please test macOS

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/499
https://github.com/apple/swift-stress-tester/pull/180
https://github.com/apple/swift-driver/pull/1140

@swift-ci please test macOS

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/499
https://github.com/apple/swift-stress-tester/pull/180
https://github.com/apple/swift-driver/pull/1140

@swift-ci please test macOS

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/499
https://github.com/apple/swift-stress-tester/pull/180

@swift-ci please test macOS

**Review by ahoppen — APPROVED** [review]

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/499
https://github.com/apple/swift-stress-tester/pull/180

@swift-ci please test macOS

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/499
https://github.com/apple/swift-stress-tester/pull/180

@swift-ci please test macOS

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/499
https://github.com/apple/swift-stress-tester/pull/180

@swift-ci please test macOS

**Comment by DougGregor** [comment]
Different spurious failure, this time from LLDB. Trying again...

**Comment by DougGregor** [comment]
https://github.com/apple/swift-syntax/pull/499
https://github.com/apple/swift-stress-tester/pull/180

@swift-ci please test macOS
